Add a DevHub manifest

chrisvanpatten opened the pull request:
**Not to merge until core Trac tickets are solved**

* https://meta.trac.wordpress.org/ticket/4388
* https://meta.trac.wordpress.org/ticket/4244

## Description

Based on [today's WordPress #docs meeting](https://wordpress.slack.com/archives/C02RP4WU5/p1556552374076400) (registration required), this PR adds a new manifest intended for consumption by [DevHub](https://developer.wordpress.org).

This is intended to accomplish a few things:

- [x] Leaves the [legacy Gutenberg handbook](https://wordpress.org/gutenberg/handbook/) unchanged
- [x] Does not actually move any documentation files
- [x] Moves designer/developer handbook content to the top level in DevHub
- [x] Preserves the current location of the contributor handbook, now living as a section within the developer handbook

The plan from here would be to…

1. Re-build DevHub from this new manifest
2. Add wildcard redirects:
    ```
    https://wordpress.org/gutenberg/handbook/designers-developers/* → https://developer.wordpress.org/block-editor/*
    https://wordpress.org/gutenberg/handbook/contributors/* → https://developer.wordpress.org/block-editor/contributors/*
    ```
3. Disable manifest syncs at the legacy handbook
4. Move the contents of `./docs/designers-developers` to the top level of `./docs` and update `toc.json` accordingly (not doing this here so we can avoid unnecessarily complicating this PR)
5. Eventually, move `manifest-devhub.json` back to just `manifest.json` (once the above steps are working as expected) and consolidate down to a single build targeted at DevHub.

There may be a further desire to move the packages documentation up to the top level, but I'm avoiding that in this initial PR. I think it's worth seeing how this works first before we go further.

cc @kenshino and @coffee2code

chrisvanpatten committed:
Add a devhub manifest

chrisvanpatten commented:
We may also want to look at coordinating with the DevHub team to see if we can have some kind of global notice at `developer.wordpress.org/block-editor` saying something like

> This documentation is for the version of the block editor that's bundled with WordPress core. To view the documentation for the latest version of the Gutenberg plugin, visit the [Gutenberg GitHub repo](#).

chrisvanpatten commented:
Another improvement here might be to allow passing `--toc` and `--manifest` flags to `npm run docs:build` so we can more easily have multiple TOCs building to multiple manifests.

That said, having multiple manifests is hopefully just temporary… so maybe not worth the energy :)

mkaz reviewed (approved):
Looks good! New manifest file is created and looks to be correct.

One minor thing, after running docs:build there appeared to be a white-space difference in the file
`packages/core-data/README.md` - you will probably need to commit that or you'll get an error

nosolosw commented:
With the changes here, the top-level TOC will be:

- Designers & Developers Handbook
- Key Concepts
- Developer Documentation *
- Data Module Reference *
- Packages *
- Components *
- Designer Documentation *
- Contributors Guide *

_*_ means it contains subsections.

I'd change a few titles (the H1 in the proper README) and would move the reference docs to the bottom so it becomes:

- Gutenberg Handbook
- Key Concepts
- Developer Documentation *
- Designer Documentation *
- Contributors Documentation *
- Data Module Reference *
- Packages Reference *
- Components Reference *

I'd like to see the other reference docs also moved to the top-level for coherence (Block API Reference, Filter Reference & Tutorials), but I'm also open to use this as a base and review the rest before launch.

nosolosw commented:
A few other logistic things:

- If we merge this, the current handbook TOC will also be modified. I guess we need to provide a `toc-devhub.json` and `manifest-devhub.json` for testing in DevHub as a first step (assuming the sync DevHub process is the same).

- Note that the "previous" / "next" links at the bottom of each page are generated from the manifest order (the current handbook has [some issues](https://github.com/WordPress/gutenberg/issues/14794)). I've noticed that the `manifest-devhub.json` doesn't follow the `toc.json` section order either, so we need to fix that as well.

chrisvanpatten commented:
@nosolosw The handbook and DevHub don't directly reference `toc.json`; it's just used as a way to map out how the manifest file should be built.

Because we're changing the output of the manifest generation script to create a `manifest-devhub.json`, the current `manifest.json`, and thus the legacy handbook, should remain totally unchanged and effectively frozen. We could provide a second TOC but that would assume we want to change the legacy manifest going forward, which I don't believe we actually want.

Once the redirects are in place, we'll be able to disable new builds on the legacy handbook site entirely so the `manifest-devhub.json` can be moved so `manifest.json` is again the canonical manifest, but _only_ consumed by DevHub.

I didn't do a super deep comparison on `toc.json` and `manifest-devhub.json` but on a quick skim it seems to all be in the correct order. Is that not actually true? Are there any specific things that are out of order?

chrisvanpatten commented:
(Also my hope is that we don't need to merge this in to actually test the DevHub sync; I wrote the PR this way with the assumption that we would only need to merge it in once we were confident this was the correct path forward.)

nosolosw commented:
>  The handbook and DevHub don't directly reference toc.json; it's just used as a way to map out how the manifest file should be built.

Ah, great, that simplifies things a bit (I actually thought the sidebar was created from it).

> Also my hope is that we don't need to merge this in to actually test the DevHub sync;

Got it. Wasn't clear if we intended to merge this as it is.

> I didn't do a super deep comparison on toc.json and manifest-devhub.json but on a quick skim it seems to all be in the correct order. Is that not actually true? Are there any specific things that are out of order?

The problem wasn't the discrepancy between `toc` and `manifest` as I hinted earlier, but pages that are generated automatically (data, packages, components): they are appended to the end of the manifest, so previous/next links will be wrong for their top-level sections and the last page of the handbook.

My old-self to the rescue :sweat_smile: https://github.com/WordPress/gutenberg/issues/14794

chrisvanpatten commented:
@nosolosw Oh shoot - those placeholder replacements aren’t working? That’s definitely a problem… I understand now. Hmm…

oandregal committed:
Move reference docs to bottom

oandregal committed:
Extract tutorials to the top-level

oandregal committed:
Rename Designers & Developers Handbook to Block Editor Handbook

oandregal committed:
Rename Contributors Guide to Contributor Documentation

oandregal committed:
Rename Components to Component Reference

oandregal committed:
Rename Packages to Package Reference

nosolosw commented:
As per [this Slack convo](https://wordpress.slack.com/archives/C02RRH4GA/p1556784388011100) we agreed to apply [these changes](https://github.com/WordPress/gutenberg/pull/15254#issuecomment-487909876) and promote tutorials as a top-level section.

Pushed changes as individual commits so we can revert as necessary.

coffee2code commented:
The Block Editor handbook on DevHub has been updated to reflect these latest changes.

A request and a suggestion:

* Please update the slug for "Block Editor Handbook" to "handbook" (instead of "designers-developers"). The page with that slug is treated as the front page of the handbook, and was previously used by the "Gutenberg Handbook" page until it was omitted from the DevHub-specific manifest in favor of having "Block Editor Handbook" be the front page. (DevHub currently has two versions of the page since I manually changed the slug for an instance of that page to ensure there was a front page.)
* A number of pages still use "Gutenberg" in their titles (but not their slugs), so "Gutenberg" could be omitted or changed to "Block Editor".

oandregal committed:
Substitute old top-level README

The front-page of the handbook is the one whose slug is handbook.
This is generated by looking at the path (docs/readme.md),
so we substitute the old README with the new one.

nosolosw commented:
> Please update the slug for "Block Editor Handbook" to "handbook" (instead of "designers-developers")

Done at 740bd75 .

> A number of pages still use "Gutenberg" in their titles (but not their slugs), so "Gutenberg" could be omitted or changed to "Block Editor".

I can prepare this change independently, as to have a good signal-to-noise balance in this PR.

oandregal committed:
Add a reference to contributors as well

nosolosw commented:
> A number of pages still use "Gutenberg" in their titles (but not their slugs), so "Gutenberg" could be omitted or changed to "Block Editor".

Here's a WIP draft PR to address this https://github.com/WordPress/gutenberg/pull/15411

AFAIK the current thinking of project members is that _Gutenberg_ should be used for anything project-related and _Block Editor_ is for the first product that came of _Gutenberg_. It's a transitional period, so things may be a bit confusing and intertwinned.

youknowriad commented on docs/manifest-devhub.json:
```diff
@@ -0,0 +1,1322 @@
+	{
+		"title": "Block Editor Handbook",
+		"slug": "handbook",
+		"markdown_source": "https://raw.githubusercontent.com/WordPress/gutenberg/master/docs/readme.md",
```
Would be good to use relative links from the start to avoid the issue we had in the handbook.

chrisvanpatten replied:
@youknowriad I'm not sure I understand here? The new DevHub has been ingesting absolute URLs without any issues. What was the problem that happened in the handbook?

youknowriad replied:
We want to stop targetting master at some point and have the docs use the stable version instead. If we use relative links, we should be able to use any branch from the consumer side.

Related #14303 #14739 and https://meta.trac.wordpress.org/ticket/4244

chrisvanpatten replied:
Ahhhh, that makes perfect sense. @coffee2code any opinions/thoughts on this?

nosolosw replied:
I can bring the changes at https://github.com/WordPress/gutenberg/pull/14739 to this PR if trac/4244 is being worked on; otherwise, I'd both separately as to not block this one.

youknowriad replied:
I don't want to block the PR either, I just think this is something we should do for both the handbook and devhub.

nosolosw replied:
@coffee2code mentioned he'll work on the relative paths alongside this, so I've ported the changes in cc3ed5e

@youknowriad @chrisvanpatten I'm a bit confused about how do we plan to update the handbook going forward and wanted to flag two separate things that were asked to the meta team:

- Riad mentioned to use relative URLs. I understand this is to pinpoint the handbook to a specific branch (for example, latest WordPress release), and any updates to that branch will get reflected in the handbook (for example: in case we missed something in the docs, or want to expand them). Is that correct?

- Chris suggested to not sync the handbook on GitHub changes. I'm not sure how this would work or how we would trigger an update. Some clarity around this would be great.

I'm not familiar with the meta code that supports this, but wanted to flag the two things we're asking in case they conflict with each other.

youknowriad replied:
> (for example: in case we missed something in the docs, or want to expand them). Is that correct?

It's more related to the fact that the docs in master contain unreleased APIs and people shouldn't be reading these docs until WP releases a stable API (A version selector is better though)

> Chris suggested to not sync the handbook on GitHub changes.

I don't have opinions on the workflow itself when/how/where to sync docs. I just think it should be based on relative URLs and could target any branch from the consuming side.

chrisvanpatten replied:
We don't have a plan specifically on how/when the updates will occur, but I do generally agree that being able to do a sync of a specific Git commit hash would be best, so we can specifically sync the documentation from the specific version of Gutenberg that is pulled into Core.

I expect—at least for the time being—this would be a manual process, and would likely mean opening a new Meta ticket with each core release providing the commit reference that will need to be synced in.

chrisvanpatten replied:
(As long as we know that @coffee2code is handling building in relative link support in the short term, and we are comfortable shipping relative links on our side, then I see no reason not to make that change now, even if we don't have an exact process on how that sync will be triggered.)

chrisvanpatten replied:
(That is to say, we know that _for now_ it will need to be Scott triggering a manual sync for us; ideally we can get a more scaleable process in place for the future but this should be good enough to get us through at least WP 5.3 while we strategize on a better approach.)

nosolosw replied:
That sounds good, thanks to both for clarifying. Added a note in the issue description linking the two core Trac tickets related to this one.

nosolosw replied:
Now that I've got better context and understand how relative paths will work [in practice](https://github.com/WordPress/gutenberg/pull/14739#issuecomment-491192067) and see the [concrete steps](https://github.com/WordPress/gutenberg/pull/15254#issuecomment-491211687), I've got a question: do we actually need to stop syncing content from GitHub? Why not to pinpoint the handbook contents to a specific branch? See [Slack conversation](https://wordpress.slack.com/archives/C02QB2JS7/p1557475613027000).

If we pinpointed the handbook to `wp/5.2` (just an example, needs to be discussed), the handbook will be effectively frozen. We would have a lightweight way to update the docs if needed, though, without having to update the manifest URL in DevHub. Less work for everyone: the branch used for generating the handbook would only change once per WordPress release.

coffee2code replied:
I would recommend pinning to a stable branch, e.g. `wp/stable` or `wp/devhub` rather than to a version-specific branch. That way we can reconfigure the Block Editor handbook to point to this new branch's manifest and leave it as is. Then it would just be a matter of someone on this end of things updating the branch whenever something needs changed.

The handbooks importer (for this and other handbooks) has cron-based auto-updates, which could be modified to support being disabled, but it'd just be easier all around to leave it in place (maybe with a longer interval).

BTW, the handbook syncing is automated. Any delays in seeing content changes reflected have stemmed from the need to make updates or fixes to importer to accommodate this new handbook.

oandregal committed:
Delete unused file that wasnt removed on rebase

oandregal committed:
Use manifest-devhub instead of old manifest

oandregal committed:
Fix toc.json from rebase

oandregal committed:
Updame manifest-devhub.json

nosolosw commented:
This has been rebased with the latest changes done to the doc tool.

oandregal committed:
Use relative paths to docs

coffee2code commented:
Can I get some clarification on how relative paths were envisioned to work? Currently, two manifests are utilized.

manifest.json has:
* Location: https://raw.githubusercontent.com/WordPress/gutenberg/master/docs/manifest.json
* **OLD markdown_source example:**<br>`https://raw.githubusercontent.com/WordPress/gutenberg/master/docs/designers-developers/key-concepts.md`
* **NEW markdown_source example:**<br>`../docs/designers-developers/key-concepts.md`

manifest-devhub.json has:
* Location: https://raw.githubusercontent.com/WordPress/gutenberg/add/devhub-manifest/docs/manifest-devhub.json
* **OLD markdown_source example:**<br>`https://raw.githubusercontent.com/WordPress/gutenberg/master/docs/designers-developers/key-concepts.md`
* **NEW markdown_source example:**<br> `../docs/designers-developers/key-concepts.md`

In both cases, it appears `..` represents `https://raw.githubusercontent.com/WordPress/gutenberg/master`, assuming both manifests are meant to point to the same file (as they did before this change).

My question: for manifest-devhub.json, should the relative path resolve to:
1. `https://raw.githubusercontent.com/WordPress/gutenberg/master/docs/designers-developers/key-concepts.md`

or should it resolve to:

2. `https://raw.githubusercontent.com/WordPress/gutenberg/add/devhub-manifest/docs/designers-developers/key-concepts.md`

Bear in mind, the path that relative paths are relative to is the location for the manifest. So my belief is that (2) is the intent. This allows for a manifest in a branch to easily reference files relative to its own branch. The markdown_source values of `designers-developers/key-concepts.md`, `/designers-developers/key-concepts.md`, and `../docs/designers-developers/key-concepts.md` would all be equivalent and resolve to (2). (I have this implemented locally.)

However, if you expect the path to be relative to the repo root, then I believe you'd have to use a markdown_source value like `master/docs/designers-developers/key-concepts.md` to refer to the master branch's version of a file, or something like `add/devhub-manifest/docs/designers-developers/key-concepts.md` to refer to the version of that file in a branch/PR.

(I'm running under the assumption that these or future manifests may want to refer to non-master versions of files.)

nosolosw commented:
Comment spun from [this conversation](https://github.com/WordPress/gutenberg/pull/14739#issuecomment-491192067) with @coffee2code that I wanted to post here so people following this can participate.

Now that we're closer to the merge status, I'd like to expand on the plan to roll this:

- Re-build DevHub from this new manifest. It uses this manifest: `https://raw.githubusercontent.com/WordPress/gutenberg/ADD/DEVHUB-MANIFEST/docs/manifest-devhub.json`. Note that, when relative paths are working, content is to be pulled from this branch, not `master` (uppercase used for effect). See [Scott comment](https://github.com/WordPress/gutenberg/pull/14739#issuecomment-491192067). 
- Publish existing DevHub content.
- Add wildcard redirects from the old handbook to DevHub.
- Merge this branch, but **DO NOT DELETE IT**. It will be still used to generate the DevHub content and bad things :tm: may happen if it no longer exists.
- Change DevHub to use the Gutenberg `master` branch instead: `https://raw.githubusercontent.com/WordPress/gutenberg/MASTER/docs/manifest-devhub.json`
- Now it's safe to delete this branch.
- We can now publicly talk and share about the DevHub migration, etc.

coffee2code commented:
What's the overall ongoing plan for the handbook at https://wordpress.org/gutenberg/? That handbook uses a custom importer that differs from the one used on DevHub With the relative link changes [applied](https://github.com/WordPress/gutenberg/pull/14739/commits/5798ecb09bf0158f3310e0fec1e556d6c7d41f3f) to its manifest (`manifest.json`), the handbook content won't update unless I port relative path support to it separately.

Based on the latest comment, it looks like the entire handbook will soon redirect to the Block Editor handbook on DevHub, in which case I would be inclined to _not_ make an effort to update the importer for the legacy Gutenberg handbook. (The legacy handbook will remain as you see it today until redirects are activated, at which point the handbook will no longer be accessible.)

chrisvanpatten commented:
@coffee2code Yes, the plan is to fully deprecate the custom handbook and move into DevHub permanently! I agree; the old handbook should not be updated at this point, except to disable syncing and add redirects.

nosolosw commented:
Cross-posting here from https://meta.trac.wordpress.org/ticket/4388#comment:22 for latest updates:

1. we'll get this branch merged so the manifest can come from master. **DO NOT DELETE THIS BRANCH YET**. (@nosolosw / @chrisvanpatten)
1.1. DevHub is updated to use the manifest-devhub in `master`.
1.2. This branch can be deleted at this point.
2. handbook is rebuilt, and presumably goes live after a final sanity check, temporarily there are two handbooks
3. redirects get enabled (@coffee2code)
4. some indeterminate testing period just to make sure we're all happy, updates are successful, etc.
5. handbook gets updated to point to wp/{version} branch

oandregal reviewed (approved):
This is ready! Let ship this to users :)

nosolosw commented:
Rebased this from `master` to force Travis to do its checks dance. For some reason, it didn't when I changed this PR from "draft" to "ready for review".

nosolosw commented on docs/designers-developers/readme.md:
```diff
@@ -1,11 +0,0 @@
-# Designer & Developer Handbook
-
-The Gutenberg project is transforming the way content is created on WordPress. A block editor was the first product launched creating a new methodology for working with content. This handbook provides documentation for how designers and developers can extend the editor.
```
Note that the current handbook sync [has been disabled](https://meta.trac.wordpress.org/ticket/4244#comment:7) so this removal won't trigger any update (nor the other minor content updates).

nosolosw commented:
DO NOT DELETE THIS BRANCH YET. [See](https://github.com/WordPress/gutenberg/pull/15254#issuecomment-492115337).

chrisvanpatten commented:
Thanks for seeing this through to the finish line @nosolosw — gold stars for you! ⭐️⭐️⭐️⭐️⭐️

nosolosw commented:
For reference: @aduth locked this branch from being deleted. We should remove the lock when we're ready to proceed with the next steps. I'll post here.

coffee2code commented:
The manifest location for the new Block Editor handbook now [uses the file in master](https://meta.trac.wordpress.org/changeset/8790) and is now [available to all visitors](https://meta.trac.wordpress.org/changeset/8791). The legacy handbook now [redirects to the new handbook](https://meta.trac.wordpress.org/changeset/8793).

It should be safe to delete this branch. Then whenever it is so desired, a meta.trac ticket can be opened to re-point the handbook to a manifest file in a wp/{version} branch.

gziolo commented:
See #15785. New Block Editor handbook should be using `wp/5.2` branch which contains the snapshot of docs at the time WordPress 5.2 was released. Otherwise, it will cause more confusion.

gziolo commented:
> 5\. Eventually, move `manifest-devhub.json` back to just `manifest.json` (once the above steps are working as expected) and consolidate down to a single build targeted at DevHub.

We still keep two files in the repository but it looks like we only update `manifest-devhub.json`. Can we remove the other one?
